### DISCUSSION TITLE
Support RS256 JWT Signing

### DISCUSSION THREAD

**Discussion by ecchochan** [supabase/supabase#12759] [Feature Requests] [answered]
Update 17th July 2025: This was launched on Monday. https://supabase.com/blog/jwt-signing-keys

# Summary
[summary]: #summary

Currently Supabase Auth only allows signing JWT with HS256. Allowing RS256 would enable the possibility of integrating with 3rd party services, e.g. vendor services to verify the JWT token.

# Motivation
[motivation]: #motivation

- Integration with vendor services that require validating JWT issued from my Supabase
- Integration with [web3auth](https://web3auth.io/docs/auth-provider-setup/byo-jwt-providers)

# Plan
[plan]: #plan

- [x] Support RS256 JWT signing/verification in [supabase/gotrue](https://github.com/supabase/gotrue) 
  - [x] https://github.com/supabase/gotrue/pull/985 
  - [ ] ~Waiting for PR review & merge ❌~
  - [x] Addressed [security concerns](https://github.com/orgs/supabase/discussions/12759#discussioncomment-5995098) at https://github.com/ecchochan/gotrue/pull/2
  - [ ] PR review & merge
- [x] Support `RS256` JWT signing/verification in [supabase/storage-api](https://github.com/supabase/storage-api/blob/5380325293cfcef49a2489aaca43430da9933b79/src/auth/jwt.ts)
  - Already supporting RS256 by setting :
    - `PGRST_JWT_ALGORITHM` to `RS256`
    - `PGRST_JWT_SECRET` to the PEM private key (e.g. `"-----BEGIN PRIVATE KEY-----\nMIGH..."`)
- [x] Setup examples to use `RS256` pubkey for JWT verification in [PostgREST](https://github.com/PostgREST/postgrest)
  - Already supporting `RS256`
    -  (https://github.com/PostgREST/postgrest/discussions/2096)
- [x] Support `RS256` JWT signing/verification in [supabase/realtime](https://github.com/supabase/realtime)
  - [x] https://github.com/supabase/realtime/pull/531
  - [ ] Waiting for PR review & merge

# Generation of private key and public key using RS256

The below bash code will write to `private_key.pem` and `public_key.pem`

```bash
openssl genrsa -out private_key.pem 2048
openssl rsa -in private_key.pem -pubout -out public_key.pem
```

# JWK required for `PostgREST`

The below js code will write to `jwks.json`

```ts
// JWT_KID=your-app-id JWT_ALG=RS256 yarn tsx keys/genJwks.ts
import * as jose from 'jose';
import { readFileSync, writeFileSync } from 'node:fs';
import { resolve } from 'node:path';

if (!process.env.JWT_KID) throw new Error('JWT_KID is empty');
if (!process.env.JWT_ALG) throw new Error('JWT_ALG is empty');

const JWT_KID = process.env.JWT_KID;
const JWT_ALG = process.env.JWT_ALG;

const genJwks = async () => {
  const publicPem = readFileSync(resolve(__dirname, './public.pem'), { encoding: 'utf-8' });
  const publicKey = await jose.importSPKI(publicPem, JWT_ALG, { extractable: true });
  const exportedKey = await jose.exportJWK(publicKey);
  const key = {
    ...exportedKey,
    use: 'sig',
    alg: JWT_ALG,
    kid: JWT_KID,
  };

  const jwk = {
    keys: [key],
  };
  writeFileSync(resolve(__dirname, 'jwks.json'), JSON.stringify({ keys: [jwk] }));
};

genJwks()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });

```

**Comment by Lan-Hekary** [comment]
Hello @ecchochan ,
Thanks for the Feature Requests and the PRs that implement it.
However, I am concerned about key confusion attacks.
This link shows an exploit that arise from using BOTHHMAC algorithms and the Public Key algorithms at the same time.
https://github.com/ticarpi/jwt_tool/wiki/Known-Exploits-and-Attacks#cve-2016-5431---key-confusion-attack

I did not go through the PRs to verify how you are implementing the support for RS256, but it should disable HS256 if enabled, otherwise it will be a security risk !!

Please make sure that you read the above link and avoid these exploits.
And Revise [#985](https://github.com/supabase/gotrue/pull/985) and [#531](https://github.com/supabase/realtime/pull/531) accordingly.
Thanks.

**Comment by borjadotai** [comment]
Amazing to hear this is finally coming to supabase and all the progress already made! Any timeline on when we'll be able to start using JWKS? 🙏

**Comment by magick93** [comment]
This is a very exciting development and opens a lot of possibilities. 

Are there any updates to this?

**Comment by 9M6** [comment]
Any update on this?

**Comment by Yorkemartin** [comment]
would love to see an update on this - this specific thing looks like its going to force me to abandon supabase for my app

**Comment by vichustephen** [comment]
Hi people! Any updates on this ? I badly require this so i can uses supabase for my future projects ?

**Comment by ollebergkvist** [comment]
This is crucial for us as well, we need to be able to rotate JWTs in order to safely integrate BankID with Supabase (the most used Oauth Provider in Sweden...)

**Comment by chrisciszak** [comment]
Wow shocked that this hasnt been implemented yet. No external provider can verify my JWTs!

**Comment by 9M6** [comment]
More than 1 year and no progress on this. Sad.

**Comment by j4w8n** [comment]
No public timeline, but asymmetric JWTs are coming soon: https://github.com/supabase/auth-js/issues/873#issuecomment-2092401090

**Comment by tomekit** [comment]
I think once assymetric JWTs are implemented it would further open up ecosystem and allow external integrations. 
At the moment with symmetric JWTs external integrations are either not possible or insecure (due to secret sharing).

**Comment by odysseyofmayhem** [comment]
Is there any movement on this?  Would really be good to know if this is a planned feature.

**Reply by odysseyofmayhem** [reply]
Some more googling found this - looks like is "almost there" at was due for release Q4 2024:
https://github.com/orgs/supabase/discussions/29289

**Reply by j4w8n** [reply]
Also...https://github.com/orgs/supabase/discussions/29260

**Comment by hf** [comment, marked as answer]
Launched on 14th July 2025: https://supabase.com/blog/jwt-signing-keys
